sound/soc/sof/sof-priv.h

```diff
@@ -326,6 +332,7 @@ struct snd_sof_dev {
 	u64 irq_status;
 	int ipc_irq;
 	u32 next_comp_id; /* monotonic - reset during S3 */
+	u32 disable_ipc_queue; /* disables further queueing of ipc's */
```

Comment:
Sorry, I can't imagine why we need this, if snd_sof_ipc_free() is called, that means the whole sof driver is unloaded/removed(sof_remove()), then the checking for this flag won't actually run.

And, we only disable but never enable it?

Reply:
@keyonjie it is disabled only when the sof device is removed. When it is inserted again, this will be reset to 0 which is enabling it again.

Reply:
> Sorry, I can't imagine why we need this, if snd_sof_ipc_free() is called, that means the whole sof driver is unloaded/removed(sof_remove()), then the checking for this flag won't actually run.
@keyonjie this is in response to a comment from Mark Brown saying that we should stop any new messages being processed to prevent race condition on removal.

> 
> And, we only disable but never enable it?

Reply:
Can't we simply hold the ipc lock and then disable ipc to prevent race condition, just like what we do in hda_suspend()?
Adding a flag to the root snd_sof_dev structure only for the one time use at removal looks wasteful to me.

Reply:
@keyonjie that will work for ipc tx but ipc rx is not in the ipc_lock context.

Reply:
@ranj063 for module unload, we should do symmetric things in sof_remove(e.g. we call snd_sof_probe() in sof_probe(), so we should call snd_sof_remove() in sof_remove also, and doing this ipc disabling inside snd_sof_remove()), this looks legitimate to me, isn't it?

Reply:
@ranj063 so here doing ipc disable in snd_sof_remove() looks too late. can we add an ops callback(e.g. .ipc_disable) which actually call hda_dsp_ipc_int_disable() on APL, and then call this .ipc_disable at the beginning of snd_sof_ipc_free()?

Those .ipc_enable/disable callbacks are legitimate, and are simple enough to toggle IPC on/off in sof/ipc.c also.

Reply:
@keyonjie thats the whole point of this disable_ipc_queue flag, to avoid that race that you think is possible. With this flag set, you will not encounter it. I dont think we need to add another op for it anymore.